File: src/main/python/AoC2019_13.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 13
#

import os
from enum import Enum
from typing import Callable

import aocd

from aoc import my_aocd
from aoc.intcode import IntCode

Tile = tuple[int, int, int]


class TileId(bytes, Enum):

    def __new__(cls, id, char):
        obj = bytes.__new__(cls, [id])
        obj._value_ = id
        obj.char = char
        return obj

    EMPTY = (0, " ")
    WALL = (1, "█")
    BLOCK = (2, "▦")
    PADDLE = (3, "▀")
    BALL = (4, "◯")


def _parse(inputs: tuple[str]) -> list[int]:
    return [int(_) for _ in inputs[0].split(",")]


def _solve(ints: list[int], inp: int) -> int:
    IntCode(ints).run([inp], out := [])
    return out[-1]


def part_1(inputs: tuple[str]) -> int:
    IntCode(_parse(inputs)).run([], out := [])
    return sum(1 for o in [out[i] for i in range(2, len(out), 3)] if o == 2)


def _play(inputs: tuple[str], on_step: Callable[[Tile], None] = None) -> int:
    prog = _parse(inputs)
    prog[0] = 2
    int_code = IntCode(prog)
    inp = out = buf = []
    ball = paddle = score = None
    while True:
        int_code.run_till_has_output(inp, out)
        if int_code.halted:
            break
        buf.append(out.pop())
        if len(buf) < 3:
            continue
        x, y, id = buf
        if on_step is not None:
            on_step((x, y, id))
        buf.clear()
        if x == -1 and y == 0:
            score = id
        elif id == TileId.PADDLE.value:
            paddle = x
        elif id == TileId.BALL.value:
            ball = x
        if ball and paddle:
            inp.append(-1 if ball < paddle else 1 if ball > paddle else 0)
            ball = None
    return score


def part_2(inputs: tuple[str]) -> int:
    return _play(inputs)


def visualize_part_2(inputs: tuple[str]) -> None:
    tiles = dict[tuple[int, int], TileId]()
    score = 0

    def on_step(tile: Tile) -> None:
        nonlocal score
        os.system("cls" if os.name in ("nt", "dos") else "clear")  # nosec
        if tile[0] == -1 and tile[1] == 0:
            score = tile[2]
        else:
            try:
                tiles[(tile[0], tile[1])] = TileId(tile[2])
            except ValueError:
                pass
        for y in range(20):
            line = ""
            for x in range(35):
                if (x, y) not in tiles:
                    line += TileId.EMPTY.char
                    continue
                line += tiles[(x, y)].char
            print(line)
        print(f"Score: {score}")

    _play(inputs, on_step)


def main() -> None:
    puzzle = aocd.models.Puzzle(2019, 13)
    my_aocd.print_header(puzzle.year, puzzle.day)

    inputs = my_aocd.get_input_data(puzzle, 1)
    assert visualize_part_2(inputs) is None
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2019_13.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static java.util.stream.Collectors.toList;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

import com.github.pareronia.aoc.intcode.IntCode;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2019_13 extends AoCBase {
    
    private final List<Long> program;
    
    private AoC2019_13(final List<String> input, final boolean debug) {
        super(debug);
        assert input.size() == 1;
        this.program = IntCode.parse(input.get(0));
    }

    public static AoC2019_13 create(final List<String> input) {
        return new AoC2019_13(input, false);
    }

    public static AoC2019_13 createDebug(final List<String> input) {
        return new AoC2019_13(input, true);
    }
    
    @Override
    public Long solvePart1() {
        final IntCode intCode = new IntCode(this.program, this.debug);
        final Deque<Long> input = new ArrayDeque<>();
        final Deque<Long> output = new ArrayDeque<>();
        intCode.run(input, output);
        final List<Long> list = output.stream().collect(toList());
        return range(2, list.size(), 3).intStream()
                .mapToLong(list::get)
                .mapToObj(TileId::from)
                .filter(TileId.BLOCK::equals)
                .count();
   }
    
    @Override
    public Long solvePart2() {
        final List<Long> quarters = new ArrayList<>();
        quarters.add(2L);
        quarters.addAll(this.program.subList(1, this.program.size()));
        final IntCode intCode = new IntCode(quarters, this.debug);
        final Deque<Long> input = new ArrayDeque<>();
        final Deque<Long> output = new ArrayDeque<>();
        final List<Long> buffer = new ArrayList<>();
        Long ball = null, paddle = null;
        long score = 0;
        
        while (true) {
            intCode.runTillHasOutput(input, output);
            if (intCode.isHalted()) {
                break;
            }
            buffer.add(output.pop());
            if (buffer.size() < 3) {
                continue;
            }
            
            final long x = buffer.get(0);
            final long y = buffer.get(1);
            final long id = buffer.get(2);
            log(buffer);
            buffer.clear();
            if (x == -1 && y == 0) {
                score = id;
            } else if (TileId.from(id) == TileId.PADDLE) {
                paddle = x;
            } else if (TileId.from(id) == TileId.BALL) {
                ball = x;
            }
            if (ball != null && paddle != null) {
                final long joystick = Long.compare(ball, paddle);
                log(" -> " + joystick);
                input.add(joystick);
                ball = null;
            }
        }
        return score;
    }

    public static void main(final String[] args) throws Exception {
        final Puzzle puzzle = Aocd.puzzle(2019, 13);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2019_13.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2019_13.create(inputData)::solvePart2)
        );
    }
    
    private enum TileId {
        EMPTY(0), WALL(1), BLOCK(2), PADDLE(3), BALL(4);
        
        private long value;

        TileId(final long value) {
            this.value = value;
        }
        
        public static TileId from(final long value) {
            return Arrays.stream(values())
                    .filter(v -> v.value == value)
                    .findFirst().orElseThrow();
        }
    }
}